### PR TITLE
feat(remote-build): add GitRepo class

### LINKED ISSUE #4320: remote-build: create wrapper for the pygit2 library
### What needs to get done

Create a new module `snapcraft/remote_build/git.py` to wrap the [GitPython](https://gitpython.readthedocs.io/en/stable/) library.

Add the following functions:

- `init(exist_ok: bool)`
- `is_repo()`
- `is_dirty()`
- `add(add_all: bool)`
- `commit(message: str)`
- `push(refspec: str = “HEAD”, url: str)`

### Why it needs to get done

The legacy code uses snapcraft_legacy’s [Git](https://github.com/snapcore/snapcraft/blob/423826a119074fd7776e01f405c51a753292b6cd/snapcraft_legacy/internal/sources/_git.py#L27) class for git-related operations.  In the new code, this will be replaced with the [GitPython](https://gitpython.readthedocs.io/en/stable/) library, encapsulated by a wrapper class.

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

Fixes #4320
(CRAFT-1975)

### CODE REVIEW

**Review by cmatsuoka — APPROVED** [review]
Looks great! (but maybe one test missing)

File: snapcraft/remote/git.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+"""Git repository class and helper utilities."""
+
+import logging
+from pathlib import Path
+
+import pygit2
+
+from .errors import GitError
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/remote/git.py:26]: The latter - the directory `snapcraft/remote/` is intended to be spun off to a new library.

File: snapcraft/remote/git.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+        raises GitError: if the name could not be resolved
+        """
+        try:
+            reference = self._repo.lookup_reference_dwim(ref).name
+            logger.debug("Resolved reference %r for name %r", reference, ref)
+            return reference
+        # raises a KeyError if the ref does not exist and a GitError for git errors
+        except (pygit2.GitError, KeyError) as error:
+            raise GitError(
```
Comment (sergiusens) [inline snapcraft/remote/git.py:187]: this could be a different error inheriting from GitError if you plan to treat it differently

**Review by sergiusens — APPROVED** [review]

File: tests/unit/remote/test_git.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+    with pytest.raises(GitError) as raised:
+        assert is_repo(new_dir)
+
+    assert raised.value.details == (
+        f"Could not check for git repository in {str(new_dir)!r}."
+    )
+
+
+def test_init_repo(new_dir):
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/unit/remote/test_git.py:65]: Done

**Review by cmatsuoka — APPROVED** [review]
Thanks!

File: snapcraft/remote/git.py (hunk 2 above)
Comment (mr-cal) [reply snapcraft/remote/git.py:187]: I don't believe I will handle this differently so I'll leave it as-is for now.

**Pull request merged**